contracts/MarginCalculator.sol

```diff
@@ -2,11 +2,319 @@
+     * @param _vault the theoretical vault that needs to be checked
+     * @param _denominated the token the result is denominated in. Must be the same as short.collateral for now.
+     * @return netValue the amount by which the margin is above or below the required amount.
+     * @return isExcess true if there's excess margin in the vault. In this case, collateral can be taken out from the vault. False if there is insufficient margin and additional collateral needs to be added to the vault to create the position.
+     */
+    function getExcessMargin(MarginAccount.Vault memory _vault, address _denominated)
+        public
+        view
+        returns (uint256, bool)
+    {
+        // ensure the number of collateral, long and short array is valid.
+        _checkIsValidSpread(_vault);
+        // ensure the long asset is valid for the short asset.
+        require(_isMarginableLong(_vault), "MarginCalculator: long asset not marginable for short asset");
+
+        // collateral amount is always positive.
+        FixedPointInt256.FixedPointInt memory collateralAmount = (_vault.collateralAssets.length == 0) ||
+            (_vault.collateralAssets[0] == address(0))
+            ? _uint256ToFixedPointInt(0)
+            : _uint256ToFixedPointInt(_vault.collateralAmounts[0]);
+
+        // Vault contains no short tokens: return collateral value.
+        if (_vault.shortOtokens.length == 0 || _vault.shortOtokens[0] == address(0))
+            return (SignedConverter.intToUint(collateralAmount.value), true);
+
+        // For the currenct version, ensure denominated == short.collateral
+        address shortCollateral = OtokenInterface(_vault.shortOtokens[0]).collateralAsset();
+        require(
+            shortCollateral == _denominated,
+            "MarginCalculator: Denomintated token should be the short otoken's collateral"
+        );
+
+        FixedPointInt256.FixedPointInt memory marginRequirement = _getMarginRequired(_vault);
+        // if marginRequirement > 0, the long assets cannot cover the max loss of short assets in the vault.
+        // will need to check if collateral - marginRequirement is greater than 0.
+        FixedPointInt256.FixedPointInt memory excessMargin = collateralAmount.sub(marginRequirement);
+        bool isExcess = excessMargin.isGreaterThanOrEqual(_uint256ToFixedPointInt(0));
+        uint256 netValue = SignedConverter.intToUint(excessMargin.value);
+        return (netValue, isExcess);
+    }
+
+    /**
+     * @notice Calculate the amount of collateral needed for a spread vault.
+     * @dev The vault passed in already pass amount array length = asset array length check.
+     * @param _vault the theoretical vault that needs to be checked
+     * @return marginRequired the minimal amount of collateral needed in a vault.
+     */
+    function _getMarginRequired(MarginAccount.Vault memory _vault)
+        internal
+        view
+        returns (FixedPointInt256.FixedPointInt memory marginRequired)
```

Comment:
Will be better to remove the variable assignment here `marginRequired` and do it inside function + `return`, I think it is more clear and easy to find the return statement.